### DISCUSSION TITLE
How to change the keyboard shortcuts in Algolia Search component?

### DISCUSSION THREAD

**Discussion by idontknowjs** [facebook/docusaurus#4952] [Q&A]
Swizzling the algolia search component `@docusaurus/theme-search-algolia` doesn't seem to show the Keyboard shortcuts used

**Comment by idontknowjs** [comment]
It seems like a part of the `@docsearch/react` package. Is there any way to edit this?

### Motivation behind this
The <kbd>Ctrl</kbd> + <kbd>K</kbd> key is a common key for enabling narrator view for accessibility tools., so whenever <kbd>Ctrl</kbd> + <kbd>K</kbd> keys are selected to enter focus into the search bar. The screen reader goes off.

**Reply by slorber** [reply]
Afaik this was choosen carefully by the Docsearch team, not sure keyboard shortcuts are customizable. 

cc @shortcuts can we customize the shortcuts? 🤪 😅

**Reply by shortcuts** [reply]
> cc @shortcuts can we customize the shortcuts?

(good one) Sorry, it's not possible to change the shortcuts in DocSearch. In web browsers, <kbd>cmd </kbd> + <kbd>k</kbd> is used to focus the omnibox, so we thought it was a good fit!

If this is something you'd like to add @covalentbond, you're more than welcome! Here's where the logic happen: 
https://github.com/algolia/docsearch/blob/next/packages/docsearch-react/src/useDocSearchKeyboardEvents.ts#L39-L46

Adding a new property that overrides the <kbd>k</kbd> but defaults it might be pretty straight forward!

**Reply by idontknowjs** [reply]
@slorber we at https://docs.zowe.org are using two DocSearch components. One in the navbar & another in the landing page.

The DocSearch present in the navbar seem to work as expected with `contextualSearch: "true"` defined in [docusaurus.config.js](https://github.com/zowe/docs-site/blob/master/docusaurus.config.js#L168) file.

But passing `contextualSearch = "true"` in DocSearch component used in home page seem to return the results of multiple version. Can you guide me how to fix this?
```
<DocSearch
  apiKey="59ff39ed48d0820010c7e09fc4b677bf"
  indexName="zowe"
  contextualSearch="true"
/>
```

I found a another website (https://docs.mia-platform.eu) similar to our use case.

**Reply by slorber** [reply]
contextualSearch is a feature of Docusaurus, not DocSearch: it permits to provide Algolia with the correct facet filters according to the currently selected version (what the dropdown displays).

On the homepage, you could hardcode a specific/latest version in your docsearch facet filters, or you can add contextual search there too.


We use this hook internally:

```js
import useContextualSearchFilters from '@theme/hooks/useContextualSearchFilters';

export default function useAlgoliaContextualFacetFilters() {
  const {locale, tags} = useContextualSearchFilters();
  const languageFilter = `language:${locale}`;
  const tagsFilter = tags.map((tag) => `docusaurus_tag:${tag}`);
  return [languageFilter, tagsFilter];
}
```

You can use it too on your homepage:

```js
import useAlgoliaContextualFacetFilters from '@theme/hooks/useAlgoliaContextualFacetFilters';

function Homepage() {
  const facetFilters = useAlgoliaContextualFacetFilters();
  return (
    <DocSearch
      apiKey="59ff39ed48d0820010c7e09fc4b677bf"
      indexName="zowe"
      facetFilters={facetFilters}
    />
  )
} 
```

Didn't test, but it should work.

**Reply by idontknowjs** [reply]
Hey, the code snippet provided gives this error:
```
Error: Can't find docs preferred context, maybe you forgot to use the "DocsPreferredVersionContextProvider"?
```

**Reply by idontknowjs** [reply]
Note: This is the configuration of [docusaurus.config.js](https://github.com/zowe/docs-site/blob/master/docusaurus.config.js#L168). Not sure if it is needed to make some changes here:
```
    algolia: {
      apiKey: "59ff39ed48d0820010c7e09fc4b677bf",
      indexName: "zowe",
      contextualSearch: true,
    },
```

You can test this in your [forked copy](https://github.com/slorber/zowe-test/blob/main/src/pages/index.js#L30) also if needed. :)

**Reply by slorber** [reply]
> Error: Can't find docs preferred context, maybe you forgot to use the "DocsPreferredVersionContextProvider"?

You need to use the hook under the layout component, because it's the layout that declares the provider.

`<Layout><CompUsingHook></Layout>` => an intermediate comp is required for now

**Reply by idontknowjs** [reply]
Yes, it's wrapped inside the `<Layout></Layout>` only. But still gives the same error. This is the `Home` component:
```
import useAlgoliaContextualFacetFilters from '@theme/hooks/useAlgoliaContextualFacetFilters';
import { DocSearch } from "@docsearch/react";

function Home() {
  const facetFilters = useAlgoliaContextualFacetFilters();
  return (
    <Layout description="Zowe">
      <header className={clsx("hero", styles.heroBanner)}>
          <div className="searchDiv">
              <DocSearch
              apiKey="59ff39ed48d0820010c7e09fc4b677bf"
              indexName="zowe"
              facetFilters={facetFilters}
              />
          </div>
      </header>
    </Layout>
  );
}
```

**Reply by slorber** [reply]
This is exactlly what you should NOT do :)

![image](https://user-images.githubusercontent.com/749374/125064895-6dbf0300-e0b1-11eb-9c4f-8398dde5427f.png)


Youy are using the hook in the component that render the Layout. You should use the hook in a subcomponent. Don't render HTML content directly inside Layout: create a separate component instead., like "SearchHeader"

**Reply by idontknowjs** [reply]
Tried creating a new `SearchHeader` sub-component & used the hook in the sub-component. Now the error has resolved, but the results generated from the search gives multiple versions only. Not sure, if I'm doing something wrong here. 😅

## `SearchHeader.js`
```
import React from 'react'
import { DocSearch } from "@docsearch/react";
import useAlgoliaContextualFacetFilters from '@theme/hooks/useAlgoliaContextualFacetFilters';

function SearchHeader() {
  const facetFilters = useAlgoliaContextualFacetFilters();
  return (
    <DocSearch
      apiKey="59ff39ed48d0820010c7e09fc4b677bf"
      indexName="zowe"
      facetFilters={facetFilters}
    />
  )
}

export default SearchHeader
```

## `index.js`
```
import React from 'react'
import Layout from "@theme/Layout";
import SearchHeader from "./SearchHeader";

function Home() {
  return (
    <Layout>
      <SearchHeader />
    </Layout>
  )
}

export default Home;
```

**Reply by slorber** [reply]
> the results generated from the search gives multiple versions only

I don't understand what you mean here?

Can you log the facetFilters and paste its value here?

This array contains the search filters.

Are the filters incorrect, or are you applying the filters badly to the DocSearch component?
I didn't read the doc of your DocSearch comp and we don't even use it in Docusaurus. 
Make sure to pass it the facet filters the correct way, it might as well be `searchParameters={{facetFilters}}`, I don't know.

**Reply by idontknowjs** [reply]
@slorber strangely earlier with `facetFilters={facetFilters}` the search gave result of docs from multiple versions which is not expected. 

But with `searchParameters={{facetFilters}}`. Whenever I'm running locally. The search seem to work successfully & return results from current version. But fails only while building stating the same error message.

Can you please take a look at the changes in this [commit](https://github.com/covalentbond/docs-site/commit/5378a8703ff13ab722b30743265d899cc505f473). This is probably failing in build.

### `facetFilters` log

Yes it does return an array only.
```
Array(2)
0: "language:en"
1: Array(2)
0: "docusaurus_tag:default"
1: "docusaurus_tag:docs-default-current"
length: 2
__proto__: Array(0)
length: 2
__proto__: Array(0)
```

**Comment by slorber** [comment]
It's not clear to me how it is failing. Is there an error message?

Le ven. 9 juil. 2021 à 19:55, Arijit Kundu ***@***.***> a
écrit :

**Reply by idontknowjs** [reply]
This [netlify build](https://app.netlify.com/sites/zowe-docs-master/deploys/60e886f4eaef600007d505b9) might help which failed.

**Comment by slorber** [comment]
You can't put components into pages/components like that, because
Docusaurus will try to create a page for each standalone component.

We have added a _ prefix so that JS files in pages don't lead to a page
creating

Le ven. 9 juil. 2021 à 21:50, Arijit Kundu ***@***.***> a
écrit :

**Reply by idontknowjs** [reply]
Thank you so much! The build passed & the search seem to work as well!

**Comment by slorber** [comment]
You can clearly see that Docusaurus is trying to render the comp as a page
here:

"/components/SearchHeader/SearchHeader" because of following error:
